[processing] Use native QGIS map layer combobox

nyalldawson opened the pull request:
This PR adds a bunch of functionality to QgsMapLayerComboBox which was present in processing's custom map layer combobox, and then switches processing across to use the standard map layer combox when you run an algorithm outside the toolbox.

Benefits include:
- reduction of duplicate code, removal of code with no unit testing
- proper sorting of layers
- icons for layer types
- combobox responds immediately to layer addition/removal

nyalldawson committed:
Allow not set choice in QgsMapLayerComboBox

nyalldawson committed:
Allow showing CRS in QgsMapLayerComboBox

nyalldawson committed:
Allow showing additional items in QgsMapLayerComboBox

These may represent additional layers such as layers which
are not included in the map layer registry, or paths to
layers which have not yet been loaded into QGIS.

nyalldawson committed:
Allow filtering QgsMapLayerComboBox by data provider

nyalldawson committed:
[processing] Use native QGIS map layer combobox

Switches the custom map layer combo box used in the run algorithm
dialog across to the standard QGIS map layer combo box

nirvn commented:
That's great, it'll help _a lot_ in making processing feel and act like the rest of QGIS.

nyalldawson committed:
[processing] Add file picker to vector/raster/table parameters
in modeller

Makes it obvious that users can set an algorithm vector/raster/
table input to a fixed table

3nids commented on src/core/qgsmaplayermodel.h:
```diff
@@ -32,13 +32,21 @@ class QgsMapLayer;
       LayerIdRole = Qt::UserRole + 1, //!< Stores the map layer ID
       LayerRole, //!< Stores pointer to the map layer itself
+      IsEmptyRole, //!< True if index corresponds to the empty (not set) value
+      IsAdditionalRole, //!< True if index corresponds to an additional (non map layer) item
```
For consistency, better not having `Is*` ?

3nids commented on src/core/qgsmaplayermodel.cpp:
```diff
@@ -35,6 +37,7 @@ QgsMapLayerModel::QgsMapLayerModel( QObject *parent )
     : QAbstractItemModel( parent )
     , mLayersChecked( QMap<QString, Qt::CheckState>() )
     , mItemCheckable( false )
+    , mAllowEmpty( false )
```
missing `mShowCrs` definition?

nyalldawson replied:
Thanks! That's probably why Travis was failing

3nids commented:
Cool, thanks for this!

nyalldawson committed:
Rename enum values for consistency

nyalldawson committed:
Show layer path tooltips in QgsMapLayerComboBox

Handy for differentiating between multiple layers with the
same name

nyalldawson committed:
Fix build
